### PR TITLE
compressor: always insert Vary headers for compressible resources

### LINKED ISSUE #11343: Vary header is not always set for compressible responses
*Description*:

When the compressor or gzip filter is enabled, the `vary: accept-encoding` header is only set when the response is actually compressed. This can cause unexpected behaviour in caching proxies, depending on the order of the request.

*Repro steps*:

With a regular caching proxy in front of envoy, and a cacheable & compressible response:

1. `curl host/file.txt -v --compressed` - Compressed response is returned, expected.
1. `curl host/file.txt -v` - Uncompressed response is returned, expected.

Clear proxy cache, and do the reverse:

1. `curl host/file.txt -v` - Uncompressed response is returned, expected.
1. `curl host/file.txt -v --compressed` - Uncompressed response is returned, **unexpected**!

*How to fix:*

Envoy needs to always set `vary: accept-encoding` for any compressible responses. Otherwise, if an uncompressed response is cached, the proxy won't know to fetch a new incoming request with `Accept-Encoding: gzip` from upstream.

### PR DESCRIPTION
**Pull request by rojkov**
Commit Message: compressor: always insert Vary headers for compressible resources
Additional Description: Even if we decide not to compress a response due to incompatible Accept-Encoding value, the Vary header needs to be inserted to let a caching proxy in front of Envoy know that the requested resource still can be served with compression applied.
Risk Level: low
Testing: manually tested with Squid and Polipo caching proxies in front of Envoy, ran unit tests.
Docs Changes: updated docs/root/configuration/http/http_filters/compressor_filter.rst
Release Notes: added a note to current.rst
Fixes #11343 

Co-authored-by: Gil Pedersen <git@gpost.dk>

### CODE REVIEW

File: source/extensions/filters/http/common/compressor/compressor.cc [hunk 1]
```diff
@@ -126,6 +128,14 @@ Http::FilterHeadersStatus CompressorFilter::encodeHeaders(Http::ResponseHeaderMa
   } else {
     config_->stats().not_compressed_.inc();
   }
+
+  // Even if we decided not to compress due to incompatible Accept-Encoding value,
+  // the Vary header would need to be inserted to let a caching proxy in front of Envoy
+  // know that the requested resource still can be served with compression applied.
```
Comment (dio) [inline source/extensions/filters/http/common/compressor/compressor.cc:134]: I think we should put this in `docs/root/configuration/http/http_filters/compressor_filter.rst` as well. WDYT?

File: source/extensions/filters/http/common/compressor/compressor.cc [hunk 2]
```diff
@@ -111,13 +111,15 @@ void CompressorFilter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallba
...
-      isAcceptEncodingAllowed(headers) && isContentTypeAllowed(headers) &&
-      !hasCacheControlNoTransform(headers) && isEtagAllowed(headers) &&
-      isTransferEncodingAllowed(headers) && !headers.getInline(content_encoding_handle.handle())) {
+  const bool isEnabledAndContentLengthBigEnough =
+      config_->enabled() && isMinimumContentLength(headers);
+  const bool isCompressible = isEnabledAndContentLengthBigEnough && isContentTypeAllowed(headers) &&
+                              !hasCacheControlNoTransform(headers) && isEtagAllowed(headers) &&
+                              !headers.getInline(content_encoding_handle.handle());
+  if (!end_stream && isEnabledAndContentLengthBigEnough && isAcceptEncodingAllowed(headers) &&
+      isCompressible && isTransferEncodingAllowed(headers)) {
```
Comment (dio) [inline source/extensions/filters/http/common/compressor/compressor.cc:120]: Thanks for grouping these bools!

File: source/extensions/filters/http/common/compressor/compressor.cc (hunk 1 above)
Comment (rojkov) [reply source/extensions/filters/http/common/compressor/compressor.cc:134]: Right! I overlooked the doc needs to be corrected too. Will update.

Comment (rojkov) [reply source/extensions/filters/http/common/compressor/compressor.cc:134]: @dio Updated!

**Comment by dio** [comment]
Sorry, one more thing: let's have it in https://github.com/envoyproxy/envoy/blob/master/docs/root/version_history/current.rst too.

**Comment by rojkov** [comment]
> Sorry, one more thing: let's have it in https://github.com/envoyproxy/envoy/blob/master/docs/root/version_history/current.rst too.

Sure. Added a note.

Also edited the PR's description.

File: docs/root/version_history/current.rst [hunk 3]
```diff
@@ -9,6 +9,24 @@ Minor Behavior Changes
 ----------------------
 *Changes that may cause incompatibilities for some users, but should not for most*
 
+* access loggers: applied existing buffer limits to access logs, as well as :ref:`stats <config_access_log_stats>` for logged / dropped logs. This can be reverted temporarily by setting runtime feature `envoy.reloadable_features.disallow_unbounded_access_logs` to false.
+* build: run as non-root inside Docker containers. Existing behaviour can be restored by setting the environment variable `ENVOY_UID` to `0`. `ENVOY_UID` and `ENVOY_GID` can be used to set the envoy user's `uid` and `gid` respectively.
+* compressor: always insert `Vary` headers for compressible resources even if it's decided not to compress a response due to incompatible `Accept-Encoding` value. The `Vary` header needs to be inserted to let a caching proxy in front of Envoy know that the requested resource still can be served with compression applied.
```
Comment (dio) [inline docs/root/version_history/current.rst:14]: Seems like this is the only one that we want to add?

Comment (rojkov) [reply docs/root/version_history/current.rst:12]: Oops, I shouldn't have trusted git merge blindly. Fixed.

**Comment by dio** [comment]
Hum, seems like it has conflict. Please sync with head.

**Comment by rojkov** [comment]
@dio Should be good now :)

**Comment by dio** [comment]
@ggreenway could you help to take a pass as senior maintainer for this? Thanks!

**Pull request merged**